common/rsyslog_plugin.cpp

```diff
@@ -0,0 +1,169 @@
+#include <iostream>
+#include <vector>
+#include <string>
+#include <cstdlib>
+#include <cstring>
+#include <fstream>
+#include <unordered_set>
+#include <regex>
+#include <tuple>
+#include "logger.h"
+#include "json.hpp"
+
+using namespace std;
+using namespace nlohmann;
+
+string LOG_FILE_NAME = "/tmp/{}_logfile";
+ofstream g_outfile;
+json g_regexList = json::array();
+vector<regex> g_expressions;
+unordered_set<string> g_errReportedTags;
+
+void createRegexList(json regexList) {
+	for(int i = 0; i < regexList.size(); i++) {
+		string regexString = regexList[i]["regex"];
+		regex expression(regexString);
+		g_expressions.push_back(expression);
+	}
+}
+
+void loadRegexlist(string rcPath) {
+	fstream rcfile;
+	rcfile.open(rcPath, ios::in);
+	if (!rcfile) {
+		cerr << "No such file: " << rcPath << endl;
+		return;
+	}
+	rcfile >> g_regexList;
+	createRegexList(g_regexList);
+}
+
+tuple<bool, json> matchRegex(string msg, string pgmName) {
+	bool isMatch = false;
+	json structuredEvent;
+
+	for (int i = 0; i < g_regexList.size(); i++) {
+		match_results<string::const_iterator> matchResults;
+		regex_search(msg, matchResults, g_expressions[i]);
+		vector<string> groups;
+		// first match in groups is entire msg string
+		for(int j = 1; j < matchResults.size(); j++) {
+			groups.push_back(matchResults.str(i));
+		}
+		vector<string> params = g_regexList[i]["params"];
+		string tag = g_regexList[i]["tag"];
+		structuredEvent["tag"] = tag;
+		structuredEvent["program"] = pgmName;
+		if(params.size() != groups.size() && g_errReportedTags.find(tag) == g_errReportedTags.end()) {
+			g_errReportedTags.insert(tag);
+			structuredEvent["parsed"] = groups;
+		} else {
+			for(int j = 0; j < params.size(); j++) {
+				structuredEvent[params[j]] = groups[j];
+			}
+			isMatch = true;
```

Comment:
We can take true return value from regex_search as match or not.
No need to search further if params count mismatch.
If count  does not match { raise syslog; return false }
else {return true with event }